Add pickle support to python bindings `Index` class

dbespalov opened the pull request:
## Changes 

* [Index](https://github.com/dbespalov/hnswlib/blob/python_bindings_pickle_io/python_bindings/bindings.cpp) class
  * Static factory methods `createFromIndex` and `createFromParams` for Index construction from another Index object or Index parameter tuple, respectively
  * Method `getIndexParams` serializes Index object. Returns a tuple with Index parameters.
  * Method `getAnnData` returns a tuple with hnsw-specific parameters. Copy of `appr_alg->data_level0_memory_` and `appr_alg->linkLists_` are returned as `py::array_t<char>` objects to avoid additional copying from python side

* Python bindings for [Index](https://github.com/dbespalov/hnswlib/blob/python_bindings_pickle_io/python_bindings/bindings.cpp) class 
  * Index serialization is implemented with `py::pickle` definition from pybind11
  * Bind `Index.__init__` to static factory methods `createFromIndex` and `createFromParams`
  * Expose parameters of the hnsw index as read-only properties in python:
    * `space_name, dim, max_elements, element_count, ef_construction, M, num_threads, ef`
  * And, two properties that support read and write:
    * `ef, num_threads`

* Updated API documentation and the first python example in [README.md](https://github.com/dbespalov/hnswlib/blob/python_bindings_pickle_io/README.md)

* New test in [python_bindings/tests/bindings_test_pickle.py](https://github.com/dbespalov/hnswlib/blob/python_bindings_pickle_io/python_bindings/tests/bindings_test_pickle.py).
  * Verifies that results of knn_query match results from copies of the same index. `Index` objects are copied using round-trip pickling.
  * Verify that knn_query gives recall of (almost) 100% for `k=25` using three spaces. Brute-force search is used to return ground-truth labels for the randomly generated items.
  * Create separate test for each space 
  * Sample test output:
```bash
> python3 -m unittest  tests/bindings_test_pickle.py -k Inner
Running pickle tests for <hnswlib.Index(space='ip', dim=48)>
Warning: 1 labels are missing from ann results (k=25, err_thresh=5)
Warning: 1 labels are missing from ann results (k=25, err_thresh=5)
 ... 
Warning: 16 ann distance values are different from brute-force values (total # of values=5000, dists_thresh=50)

.
----------------------------------------------------------------------
Ran 1 test in 19.057s

OK
```

yurymalkov commented:
@dbespalov 
Great! Thanks! I'll check it this week.

dbespalov commented:
@yurymalkov actually, could you hold off with the review?

I think the current implementation is pretty inefficient. `__getstate__` makes two copies during serialization (with a call to `.str()`, and when casting with `py::bytes`). And `__setstate__` makes at least one copy (when initing content of istream)

I'm going to try to take another stab at serialization logic using [pybind11/eigen ](https://pybind11.readthedocs.io/en/stable/advanced/cast/eigen.html), so it does not create unnecessary copies of index during pickling

yurymalkov commented:
@dbespalov Sure. There is no rush. 
Thank you so much for the work!

dbespalov commented:
@yurymalkov I updated the PR and it should now be ready for the review! 

Copies of `appr_alg->linkLists_` and `appr_alg->data_level0_memory_` are serialized as `array_t<char>` objects (these look like numpy arrays to python).

I also did a quick test to benchmark Index serialization runtime and memory usage against a file-based pickling logic. The following snippet creates 25 copies of Index object with 100k points in 512 dim. 

```python
import hnswlib

data = np.float32(np.random.random((100_000, 512))) # create 100k points with dim=512

ann = hnswlib.Index(space='cosine', dim=512) # create ann index with cosine distance
ann.init_index(max_elements=100_000, ef_construction=10, M=8) 

t0 = time.time();  
ann.add_items(data); # add points to index
t1=time.time() 

print(f"Building index took {(t1-t0)}s")

### create 25 copies of ann object using four strategies: file, memory, param, clone
for ii in range(25):
    if sys.argv[-1] == 'file':   ### use file-based pickle to create copy ann 
        copyreg.pickle(hnswlib.Index, pickle_ann) ### use temp files for ann serialization
        bb=pickle.dumps(ann)
        ann=pickle.loads(bb)
    if sys.argv[-1] == 'memory': ### copy ann with round-trip pickle
        bb=pickle.dumps(ann)
        ann=pickle.loads(bb)
   if sys.argv[-1] == 'param':  ### create new index using ann's state tuple
        ann=hnswlib.Index(ann.__getstate__())
   if sys.argv[-1] == 'clone':  ### use copy constructor to copy ann object
        ann= hnswlib.Index(ann)
        
t2=time.time()
print(f"Making 25 copies of index took {(t2-t1)}s")
```

I got the following timing results when running the snippet locally:

```bash
> python3 tests/test_ann.py file
Building index took 1.471487045288086s
pickling ann using temp file /var/folders/_4/l7lvz56s609867vwt5r5bkj98kcygg/T/tmpw3ydj4uu.ann
unpickling ann using temp file /var/folders/_4/l7lvz56s609867vwt5r5bkj98kcygg/T/tmpwu8xnto7.ann
 ...
Making 25 copies of index took 55.927014112472534s

> python3 tests/test_ann.py memory
Building index took 1.5638556480407715s
Making 25 copies of index took 20.929177284240723s

> python3 tests/test_ann.py param
Building index took 1.4085211753845215s
Making 25 copies of index took 8.599105834960938s

> python3 tests/test_ann.py clone
Building index took 1.4074420928955078s
Making 25 copies of index took 8.865073919296265s
```

The in-memory pickling (implemented in this PR) is more than 2x faster and the memory usage was comparable to the file-based method (also no leaks).

yurymalkov commented:
Thank you! I'll review it in the few following days.

yurymalkov commented on README.md:
```diff
@@ -76,14 +76,34 @@ Index methods:
+
+Properties of `hnswlib.Index` that support reading and writing:
+
+* `ef` - parameter controlling query time/accuracy trade-off. Note that setting property `p.ef` prior to index initialization with `p.init_index(...)` will raise an error. 
```
It seems like you've added a check for that condition in the, so there should be no error.

dbespalov replied:
Apologies. Forgot about that note. Removed.

yurymalkov reviewed (approved):
Thank so much for the PR! This is a very helpful feature!
I think it is ok to merge, but few things can be improved.

dbespalov commented on python_bindings/bindings.cpp:
```diff
@@ -254,6 +276,233 @@ class Index {
+      char* link_list_npy = (char *) malloc(link_npy_size);
+      int* element_levels_npy = (int *) malloc(appr_alg->element_levels_.size()*sizeof(int));
+
+      // std::unordered_map<labeltype, tableint> label_lookup_;
```
code comment removed. sorry about that!

dbespalov commented on python_bindings/bindings.cpp:
```diff
@@ -254,6 +276,233 @@ class Index {
+      new_index->seed = index_params[6].cast<size_t>();
+
+      if (index_inited_){
+        ////                      hnswlib::HierarchicalNSW<dist_t>(l2space,            maxElements,                  M,                             efConstruction,                random_seed);
```
removed. sorry about that

dbespalov commented on python_bindings/tests/bindings_test_pickle.py:
```diff
@@ -0,0 +1,159 @@
+
+class PickleUnitTests(unittest.TestCase):
+
+    def setUp(self):
```
good point! Removed camel-casing from test method names. Note that `setUp` appears to be a reserved name in unittest: https://docs.python.org/3/library/unittest.html#unittest.TestCase.setUp

dbespalov commented on python_bindings/bindings.cpp:
```diff
@@ -254,6 +276,233 @@ class Index {
+
+    void setAnnData(const py::tuple t) {
+
+      std::unique_lock <std::mutex> templock(appr_alg->global);
```
good callout, thank you! I added warning comments re: serialization thread-safety with addItems method

dbespalov commented on python_bindings/tests/bindings_test_pickle.py:
```diff
@@ -0,0 +1,159 @@
+            dists[jj,ii]=get_dist(metric, items[ii, :], query_items[jj, :])
+
+    labels = np.argsort(dists, axis=1)
+    dists = np.sort(dists, axis=1)
```
I opted for arg/sort for readability. Argpartition is much faster / shaves off log(N). I added faster alternatives as comments to both statements.

dbespalov commented on README.md:
```diff
@@ -106,6 +126,16 @@ p.set_ef(50) # ef should always be > k
 # Query dataset, k - number of closest elements (returns 2 numpy arrays)
 labels, distances = p.knn_query(data, k = 1)
+
+# Index objects support pickling:
```
I added warning re: serialization is not thread safe with add_items.

dbespalov commented on python_bindings/bindings.cpp:
```diff
@@ -254,6 +276,233 @@ class Index {
+        idx++;
+      }
+
+      memset(data_level0_npy, 0, level0_npy_size);
```
Good catch! This memset is def. not needed. removed.

dbespalov commented on python_bindings/bindings.cpp:
```diff
@@ -254,6 +275,239 @@ class Index {
+
+    py::tuple getIndexParams() const {
+        /*  TODO: serialize state of random generators appr_alg->level_generator_ and appr_alg->update_probability_generator_  */
+        /*        for full reproducibility / to avoid re-initializing generators inside Index<float>::createFromParams         */
```
Added TODO block to serialize state of random generators. Currently, the random generators are re-initialized on index deserialization. This could make index operations not (fully) reproducible.

dbespalov commented:
@yurymalkov thank you for your review! I think all your comments have been addressed. I clicked "resolved conversation" for the items that have been fully addressed no longer needs attention. 

The only item remaining is synchronization of serialization with element insertion. I've added comments to warn about thread-safety of these two operation. If you think there is a better/easier way to synchronize serialization, so that it is thread-safe with element insertion, I can take a stab at that. Frankly, I did not look into synchronization logic for too long -- I simply repurposed some code from `hnswalg.h` that was using global lock. 

One more item that is worth pointing out: I added TODO block in `python_bindings/bindings.cpp` about serialization of states for the two random generators used in `hnswalg.h`. Current implementation re-initializes random generators using `random_seed` value that is passed to init_index. I dont think this would affect functional performance of the ann index, but it can potentially cause issues when reproducibility of the ann data-structure is required. 

TLDR: Please let me know if there is a simple fix to properly synchronize serialization and element insertion. Otherwise, I think the PR looks good to be merged.

yurymalkov commented:
@dbespalov Thank you so much for the update!!
I think a good way to synchronize updates/saves/pickles is by emulating a light read-write lock in C++ code (I think read-write lock is not a part of C++11, but not sure if C++11 has advantages over C++14 nowadays) - e.g. add a flag to lock new insertions and add a counter of the number of running insertions. Though probably there are other solutions.
Anyway I think this is totally out of scope of this PR (though is important).

One other thing I forgot to comment is maybe replacing a tuple in https://github.com/nmslib/hnswlib/pull/251/files#diff-e7a40eaa3578ae348570a4917aaf8b203f34246d035b611aa0ed07252bb24ba4R345 with a dict (that would make adding/deleting fields easier) and adding a version number for rising exceptions in case of data pickled by a newer/older version of the pickling format (e.g. what happens if we add serialized seed?). 

I think changing tuple to dict can be done in the future (it is anyway likely to break the backward compatibility), but adding a version number from the start seems to be a good move. I definitely regret I didn't add it to the save format.
I can do it myself though.

dbespalov commented:
@yurymalkov I will update the PR to add serialization version to the parameter tuple. Very good call on this one! Thank you!!

Please see my other comments below 

> @dbespalov Thank you so much for the update!!
> I think a good way to synchronize updates/saves/pickles is by emulating a light read-write lock in C++ code (I think read-write lock is not a part of C++11, but not sure if C++11 has advantages over C++14 nowadays) - e.g. add a flag to lock new insertions and add a counter of the number of running insertions. Though probably there are other solutions.
> Anyway I think this is totally out of scope of this PR (though is important).

TIL: read-write lock is not part of C++11! I did not know that -- I have not touched C/C++ in more than a few years :)

Agreed that it is out of scope for this PR.  

> One other thing I forgot to comment is maybe replacing a tuple in https://github.com/nmslib/hnswlib/pull/251/files#diff-e7a40eaa3578ae348570a4917aaf8b203f34246d035b611aa0ed07252bb24ba4R345 with a dict (that would make adding/deleting fields easier) and adding a version number for rising exceptions in case of data pickled by a newer/older version of the pickling format (e.g. what happens if we add serialized seed?).
> 

The parameter tuple should definitely be a python dict! Unfortunately, I did not get to it for this PR. 
Serialization version must be included with the parameter tuple. I will update the PR.

> I think changing tuple to dict can be done in the future (it is anyway likely to break the backward compatibility), but adding a version number from the start seems to be a good move. I definitely regret I didn't add it to the save format.
> I can do it myself though.

All good! Thank you for catching this!!

dbespalov commented:
@yurymalkov I added `Index::ser_version`, a static const data member for the serialization version.  `ser_version` is the first element of parameter tuple, which has three elements in total now. The tests are passing as well.

Let me know what you think!
![image](https://user-images.githubusercontent.com/3240798/98325862-c20f6180-1fbd-11eb-9e69-b7781052221f.png)

yurymalkov commented:
Great! Thank you so much for the contribution!
I'll merge it when we fix CI.

dbespalov commented:
@yurymalkov Thank you so much for your help with improving this PR!!

Do you know what the timeframe would be for the change to be merged to master and it ends up in pypi?

yurymalkov commented:
Hi @dbespalov. Thanks again for the PR!
I am going to consider switch from tuples to dict before merging to master to avoid future compatibility issues. Maybe @dyashuni can help?

dyashuni commented:
@yurymalkov Sure. Let me see what I can do.

dbespalov commented:
@yurymalkov makes sense. In a retrospect, I should've included this change in my PR. I can probably find some time later this week to make a quick PR for this.

yurymalkov commented:
@dbespalov Cool! Thank you!

dyashuni commented:
@dbespalov Looks like with the latest changes the first python example doesn't work in the readme. There is an issue in this line of code

        print(f"Parameters passed to constructor:  space={p_copy.space}, dim={p_copy.dim}")
    AttributeError: 'hnswlib.Index' object has no attribute 'space'

The correct property of Index class is `space_name`. However `space_name` is the same as the constructor parameter `space` but called differently. I think we should align names of variables. Do you have any ideas how can we fix the issue? I can help with fixing.

dbespalov commented:
@dyashuni oops, I missed this. I meant to rename space_name to space on the python side (to be consistent with constructor's kwargs). It should be a quick fix: 

[Line 658](https://github.com/nmslib/hnswlib/blob/develop/python_bindings/bindings.cpp#L658) that now reads
`        .def_readonly("space_name", &Index<float>::space_name)
`
should be 
`        .def_readonly("space", &Index<float>::space_name)
`

dyashuni commented:
@dbespalov Thank you, I have created a pull request with the suggested quick fix.
